test/GPv2Settlement/ComputeTradeExecutions.t.sol

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8.26;
+
+import {Vm} from "forge-std/Test.sol";
+
+import {GPv2Settlement} from "src/contracts/GPv2Settlement.sol";
+import {GPv2Order, IERC20} from "src/contracts/libraries/GPv2Order.sol";
+import {GPv2Signing} from "src/contracts/mixins/GPv2Signing.sol";
+
+import {GPv2Transfer, Helper, SettlementEncoder} from "./Helper.sol";
+
+import {Order} from "test/libraries/Order.sol";
+import {Registry, TokenRegistry} from "test/libraries/encoders/TokenRegistry.sol";
+
+abstract contract BaseComputeTradeExecutions is Helper {
+    using TokenRegistry for TokenRegistry.State;
+    using TokenRegistry for Registry;
+    using SettlementEncoder for SettlementEncoder.State;
+
+    IERC20 internal sellToken;
+    IERC20 internal buyToken;
+
+    uint256 internal executedAmount = 10 ether;
+    uint256 internal sellPrice = 1;
+    uint256 internal buyPrice = 2;
+
+    function setUp() public virtual override {
+        super.setUp();
+
+        sellToken = IERC20(makeAddr("GPv2Settlement.BaseComputeTradeExecutions: sellToken"));
+        buyToken = IERC20(makeAddr("GPv2Settlement.BaseComputeTradeExecutions: buyToken"));
+    }
+
+    function defaultOrder() internal view returns (GPv2Order.Data memory) {
+        return GPv2Order.Data({
+            sellToken: sellToken,
+            buyToken: buyToken,
+            receiver: GPv2Order.RECEIVER_SAME_AS_OWNER,
+            sellAmount: 42 ether,
+            buyAmount: 13.37 ether,
+            validTo: type(uint32).max,
+            appData: bytes32(0),
+            feeAmount: 0,
+            kind: bytes32(0),
+            partiallyFillable: false,
+            sellTokenBalance: GPv2Order.BALANCE_ERC20,
+            buyTokenBalance: GPv2Order.BALANCE_ERC20
+        });
+    }
+
+    function computeSettlementForOrder(GPv2Order.Data memory order)
+        internal
+        returns (uint256 executedSellAmount, uint256 executedBuyAmount)
+    {
+        encoder.tokenRegistry.tokenRegistry().setPrice(sellToken, sellPrice);
+        encoder.tokenRegistry.tokenRegistry().setPrice(buyToken, buyPrice);
+        encoder.signEncodeTrade(vm, trader, order, domainSeparator, GPv2Signing.Scheme.Eip712, executedAmount);
+
+        SettlementEncoder.EncodedSettlement memory encoded = encoder.encode(settlement);
+        (GPv2Transfer.Data[] memory inTransfers, GPv2Transfer.Data[] memory outTransfers) =
+            settlement.computeTradeExecutionsTest(encoded.tokens, encoded.clearingPrices, encoded.trades);
+
+        // TODO: this is required due some weirdness whereby if `settlement.computeTradeExecutionsTest` reverts,
+        // for some reason it continues executing here and results in a panic with an index out of bounds error.
+        if (inTransfers.length == 1 && outTransfers.length == 1) {
```

Comment:
I see, this is because `vm.expectRevert()` prevents the function from reverting and then execution continues like nothing happened, which include assuming that the output values are valid (which are then unallocated and of size 0, causing the panic).

<details><summary>See minimal example</summary>

```solidity
contract Revert {
    function reverting() public pure returns (uint256 out) {
        out = 42;
        revert();
    }
}

contract TestRevert is Test {
    function test_reverting() public {
        Revert r = new Revert();
        vm.expectRevert();
        uint256 out = r.reverting();
        // Note: the test fails on this assertion, 0 != 42
        assertEq(out, 42);
    }
}
```
</details>

I'd say the main reason why this is unexpected is that  `vm.expectRevert` effects are far from when the actual assertion occurs. My suggestions would be either:
1. add some flag `shouldRevert` that adds `vm.expectReverts` before the call if needed and handles the output properly in the two cases, or
2. make this a bit more similar to a wrapper around `computeTradeExecutionsTest` and just return whatever it returns; then it's clearer that we shouldn't use the return values if `computeTradeExecutionsTest` is used after `expectRevert` and no surprise otherwise.

Reply:
Why not just guard as is at the moment, but make the comment clearer now that the reason is located?

Reply:
My main reason to hesitate is that it feels like a hack: what if, for example, it happened that `computeTradeExecutionsTest` didn't return transfer of size 1 on successful execution? Then the output of this function would be incorrectly replaced by the bogus values 0/0 and it would take longer to find the source of the issue. This doesn't happen in the current code but it could happen if, for example, we decided to change that function to not consider native ETH a transfer but handle it in some other way.

However, I think this would be something that only concerns this specific test file, which is small enough that a related issue probably wouldn't take too much time to debug, so I think it wouldn't be too bad to leave it as it is plus comment and maybe changing the condition to `in.length !=0 || out.length != 0` so that this is the only case that can slip through silently.